scylla/src/transport/retry_policy.rs

```diff
@@ -216,6 +216,172 @@ impl RetrySession for DefaultRetrySession {
     }
 }
 
+/// Downgrading consistency retry policy - retries with lower consistency level if it knows\
+/// that the initial CL is unreachable. Also, it behaves as [DefaultRetryPolicy] when it believes that the initial CL is reachable.
+/// Behaviour based on [DataStax Java Driver]\
+///(https://docs.datastax.com/en/drivers/java/3.11/com/datastax/driver/core/policies/DowngradingConsistencyRetryPolicy.html)
+#[derive(Debug)]
+pub struct DowngradingConsistencyRetryPolicy;
+
+impl DowngradingConsistencyRetryPolicy {
+    pub fn new() -> DowngradingConsistencyRetryPolicy {
+        DowngradingConsistencyRetryPolicy
+    }
+}
+
+impl Default for DowngradingConsistencyRetryPolicy {
+    fn default() -> DowngradingConsistencyRetryPolicy {
+        DowngradingConsistencyRetryPolicy::new()
+    }
+}
+
+impl RetryPolicy for DowngradingConsistencyRetryPolicy {
+    fn new_session(&self) -> Box<dyn RetrySession> {
+        Box::new(DowngradingConsistencyRetrySession::new())
+    }
+
+    fn clone_boxed(&self) -> Box<dyn RetryPolicy> {
+        Box::new(DowngradingConsistencyRetryPolicy)
+    }
+}
+
+pub struct DowngradingConsistencyRetrySession {
+    was_unavailable_retry: bool,
+    was_read_timeout_retry: bool,
+    was_write_timeout_retry: bool,
+}
+
+impl DowngradingConsistencyRetrySession {
+    pub fn new() -> DowngradingConsistencyRetrySession {
+        DowngradingConsistencyRetrySession {
+            was_unavailable_retry: false,
+            was_read_timeout_retry: false,
+            was_write_timeout_retry: false,
```

Comment:
question: do we really need 3 booleans here or can we just use one? Read timeout and write timeout certainly look mutually exclusive within a single query, not sure about unavailable. In any case, should the retry policy retry once per error type, or just once and for all? In the latter case we could unify the three booleans into a single parameter indicating that we already retried

Reply:
As Java driver only pays attention to retry count being 0 or more, I guess we can use just one boolean.